packages/@lwc/engine-core/src/framework/context-provider.ts

```diff
@@ -8,16 +8,15 @@ import { isUndefined, ArrayIndexOf } from '@lwc/shared';
 import { guid } from './utils';
 import { WireAdapterConstructor, ContextValue, getAdapterToken, setAdapterToken } from './wiring';
 
-type WireContextDisconnectCallback = () => void;
-type WireContextInternalProtocolCallback = (
-    newContext: ContextValue,
-    disconnectCallback: WireContextDisconnectCallback
-) => void;
+interface WireContextInternalProtocolEventDetails {
+    setNewContext(newContext: ContextValue): void;
+    provideDisconnectedCallback(disconnectCallback: () => void): void;
+}
 interface ContextConsumer {
     provide(newContext: ContextValue): void;
 }
 
-interface WireContextEvent extends CustomEvent<WireContextInternalProtocolCallback> {}
+interface WireContextEvent extends CustomEvent<WireContextInternalProtocolEventDetails> {}
```

Comment:
maybe making this an actual class that sets readonly own properties on the constructor, to prevent any modification of the event itself.